Code/Framework/AzCore/AzCore/Jobs/JobExecutor.cpp

```diff
@@ -0,0 +1,349 @@
+                }
+                m_active.store(true, AZStd::memory_order_release);
+
+                m_thread = AZStd::thread{ [this, &initSemaphore]
+                                          {
+                                              initSemaphore.release();
+                                              Run();
+                                          },
+                                          &desc };
+            }
+
+            void Join()
+            {
+                m_active.store(false, AZStd::memory_order_release);
+                m_semaphore.release();
+                m_thread.join();
+            }
+
+            void Enqueue(TypeErasedJob* job)
+            {
+                m_queue.Enqueue(job);
+
+                if (!m_busy.exchange(true))
+                {
+                    // The worker was idle prior to enqueueing the job, release the semaphore
+                    m_semaphore.release();
+                }
+            }
+
+        private:
+            void Run()
+            {
+                while (m_active)
+                {
+                    m_busy = false;
+                    m_semaphore.acquire();
+
+                    if (!m_active)
+                    {
+                        return;
+                    }
+
+                    m_busy = true;
+
+                    TypeErasedJob* job = m_queue.TryDequeue();
+                    while (job)
+                    {
+                        job->Invoke();
+                        // Decrement counts for all job successors
+                        for (size_t j = 0; j != job->m_outboundLinkCount; ++j)
+                        {
+                            TypeErasedJob* successor = job->m_graph->m_successors[job->m_successorOffset + j];
+                            if (--successor->m_dependencyCount == 0)
+                            {
+                                m_executor->Submit(*successor);
+                            }
+                        }
+
+                        job->m_graph->Release();
+                        --m_executor->m_remaining;
+
+                        job = m_queue.TryDequeue();
+                    }
+                }
+            }
+
+            AZStd::thread m_thread;
+            AZStd::atomic<bool> m_active;
+            AZStd::atomic<bool> m_busy;
+            AZStd::binary_semaphore m_semaphore;
+
+            ::AZ::JobExecutor* m_executor;
+            JobQueue m_queue;
+        };
+    } // namespace Internal
+
+    JobExecutor& JobExecutor::Instance()
+    {
+        // TODO: Create the default executor as part of a component (as in JobManagerComponent)
+        static JobExecutor executor;
+        return executor;
+    }
+
+    JobExecutor::JobExecutor(uint32_t threadCount)
+    {
+        // TODO: Configure thread count + affinity based on configuration
+        m_threadCount = threadCount == 0 ? AZStd::thread::hardware_concurrency() : threadCount;
```

Comment:
Would be great if this is configurable through the Settings Registry. This allows users to override the thread count. This can be really helpful in profiling as well.

Reply:
Will hold off on this until a future commit that actually integrates this system into part of the frame, but duly noted (initial checkin will be system + tests, without integration into any part of the rest of the engine)